icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -9,39 +16,404 @@ namespace icicle {
   public:
     CPUMerkleTreeBackend(
       const std::vector<Hash>& layer_hashes, uint64_t leaf_element_size, uint64_t output_store_min_layer = 0)
-        : MerkleTreeBackend(layer_hashes, leaf_element_size, output_store_min_layer)
+        : MerkleTreeBackend(layer_hashes, leaf_element_size, output_store_min_layer), m_tree_already_built(false)
     {
-      ICICLE_LOG_DEBUG << "in CPUMerkleTreeBackend, have " << layer_hashes.size() << " layer hashes";
-      for (const auto& layer_hash : layer_hashes) {
-        ICICLE_LOG_DEBUG << "layer hash (" << &layer_hash << "), name: " << layer_hash.name();
+      // update layers data base with the hashes
+      const int nof_layers = layer_hashes.size();
+      m_layers.resize(nof_layers);
+      m_prooned_path_size = 0;
+      m_full_path_size = 0;
+      for (int layer_idx = 0; layer_idx < nof_layers; ++layer_idx) {
+        ICICLE_ASSERT(
+          layer_idx == nof_layers - 1 ||
+          layer_hashes[layer_idx + 1].input_default_chunk_size() % layer_hashes[layer_idx].output_size() == 0)
+          << "Each layer output size must divide the above layer input size. Otherwise its not a tree.\n"
+          << "Layer " << layer_idx << " input size = " << layer_hashes[layer_idx + 1].input_default_chunk_size() << "\n"
+          << "Layer " << layer_idx + 1 << " output size = " << layer_hashes[layer_idx].output_size() << "\n";
+
+        // initialize m_layers with hashes
+        m_layers[layer_idx].m_hash = layer_hashes[layer_idx];
+
+        // Calculate m_path_size
+        if (0 < layer_idx) {
+          m_prooned_path_size +=
+            layer_hashes[layer_idx].input_default_chunk_size() - layer_hashes[layer_idx - 1].output_size();
+          m_full_path_size += layer_hashes[layer_idx].input_default_chunk_size();
+        }
       }
     }
 
-    eIcicleError build(const std::byte* leaves, uint64_t size, const MerkleTreeConfig& config) override
+    // TODO: handle size
+    eIcicleError build(const std::byte* leaves, uint64_t leaves_size, const MerkleTreeConfig& config) override
     {
-      ICICLE_LOG_DEBUG << "CPU CPUMerkleTreeBackend::build() called with " << size << " bytes of leaves";
-      return eIcicleError::SUCCESS; // TODO: Implement tree-building logic
+      TasksManager<HashTask> task_manager(get_nof_workers(config)); // Run workers.
+      ICICLE_ASSERT(!m_tree_already_built) << "Tree cannot be built more than one time";
```

Comment:
why? I think it's worth an error message telling the user to construct another tree and return an error rather than crashing.
The problem with crashing is mainly that you crash in C++ and it's really unclear in rust/Go what happened. With an error message Rust can panic and I think it's clearer for the user where it crashed and then he also sees an error message using ICICLE_LOG_ERROR << "..."

Reply:
done